bundles/io/org.eclipse.smarthome.io.net.security/src/main/java/org/eclipse/smarthome/io/net/security/internal/NetworkServerTlsProvider.java

```diff
@@ -0,0 +1,299 @@
+     * Return the current data. Used to check if certificates are still valid.
+     * May be mocked by tests.
+     *
+     * @return The current date
+     */
+    protected Date dateNow() {
+        return new Date();
+    }
+
+    public X509Certificate getSelfCertificate(String sigAlg, KeyPair key, X500Name myname, Date firstDate,
+            long validityInS, CertificateExtensions ext) throws GeneralSecurityException {
+        X509CertImpl cert;
+        Date lastDate;
+
+        try {
+            lastDate = new Date();
+            lastDate.setTime(firstDate.getTime() + validityInS * 1000);
+
+            CertificateValidity interval = new CertificateValidity(firstDate, lastDate);
+
+            X509CertInfo info = new X509CertInfo();
+            // Add all mandatory attributes
+            info.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+            info.set(X509CertInfo.SERIAL_NUMBER,
+                    new CertificateSerialNumber(new java.util.Random().nextInt() & 0x7fffffff));
+            AlgorithmId algID = AlgorithmId.get(sigAlg);
+            info.set(X509CertInfo.ALGORITHM_ID, new CertificateAlgorithmId(algID));
+            info.set(X509CertInfo.SUBJECT, myname);
+            info.set(X509CertInfo.KEY, new CertificateX509Key(key.getPublic()));
+            info.set(X509CertInfo.VALIDITY, interval);
+            info.set(X509CertInfo.ISSUER, myname);
+            if (ext != null) {
+                info.set(X509CertInfo.EXTENSIONS, ext);
+            }
+
+            cert = new X509CertImpl(info);
+            cert.sign(key.getPrivate(), sigAlg);
+
+            return cert;
+
+        } catch (IOException e) {
+            throw new CertificateEncodingException("getSelfCert: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public SSLContext createSSLContext(@Nullable String context) throws GeneralSecurityException, IOException {
+        ContextConfiguration contextConfig = configuration.getContext(context);
+
+        KeyStore ks = openKeystore(contextConfig.keystoreFilename, contextConfig);
+        boolean createSelfSigned = true;
+        if (ks.containsAlias(contextConfig.keystoreAliasTLS)) {
+            createSelfSigned = false;
+
+            // Check certificate
+            Certificate[] certificateChain = ks.getCertificateChain(contextConfig.keystoreAliasTLS);
+            if (certificateChain == null) {
+                throw new CertificateException("No certificate found");
+            }
+            try {
+                for (Certificate c : certificateChain) {
+                    ((X509Certificate) c).checkValidity(dateNow());
+                }
+            } catch (CertificateException e) {
+                // Certificate is expired or not valid yet. Create an new custom one if configuration allows this.
+                if (!configuration.autoRefreshInvalidCertificate) {
+                    throw e;
+                }
+                createSelfSigned = true;
```

Comment:
Hm, the certificate should maybe only replaced, if it was a self-signed one before, true. I didn't see the problem while implementing because I thought, who ticks the box for "auto refresh" (default off) if he/she provides an own certificate.